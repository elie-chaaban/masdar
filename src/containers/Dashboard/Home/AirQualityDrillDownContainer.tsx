import React, {useCallback, useEffect, useRef, useState} from 'react';
import {Styles} from '../../../types';
import {createStyles} from '../../../theme';
// @ts-ignore
import {fetchHourlyBuildingsCo, fetchHourlyBuildingsTemperature, fetchHourlyBuildingsHumidity} from '../../../services/dashBoard';
//@ts-ignore
import useFetch from '../../../hooks';
import {useDispatch, useSelector} from 'react-redux';
import Highcharts from 'highcharts';
import HighchartsReact from 'highcharts-react-official';
//@ts-ignore
import Loader from '../../../components/Utils/Loader';
import {Text} from '../../../components/Common/Text';
import {TextButton} from '../../../components/Common/TextButton';
// @ts-ignore
import {setIsAirQualityDrillDownOpened} from '../../../reduxStore/actions';

const webStyles: Styles = createStyles(({colors, spacing}) => ({
  container: {
    position: 'absolute',
    bottom: 0,
    left: 48,
    minWidth: '100%',
    maxWidth: '100%',
    zIndex: 9999,
    color: colors.white
  },
  overlay: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 300,
    zIndex: 999
  },
  hourlyAirQualityDrillDownTrendLine: {
    borderRadius: 0
  },
  bottomSection: {
    ...spacing.flexHorizontally,
    justifyContent: 'center',
    backgroundColor: colors.chartsCardBackground
  },
  title: {
    ...spacing.center,
    textAlign: 'center',
    width: '100%',
    marginBottom: 20,
    maxHeight: 30
  },
  switchesContainer: {
    ...spacing.centerVertically,
    minWidth: '15%'
  },
  button: {
    width: 290,
    minHeight: 57,
    maxHeight: 57,
    marginBottom: 20
  },
  chartContainer: {
    ...spacing.flexVertically,
    justifyContent: 'space-between',
    width: '100%',
    height: '100%',
    padding: 23,
    backgroundColor: colors.consumptionCardBackground,
    flex: 2
  },
  rightSection: {minWidth: '70%', minHeight: 300}
}));

const data = {
  title: {
    text: null
  },

  chart: {
    backgroundColor: null,
    color: 'white',
    height: '200'
  },
  yAxis: {
    title: {
      text: ''
    }
  },

  xAxis: {
    categories: []
  },

  legend: {
    layout: 'horizontal',
    align: 'center',
    verticalAlign: 'bottom'
  },

  plotOptions: {},

  series: []
};

const tabletStyles: Styles = createStyles(({colors, spacing}) => ({
  ...webStyles,
  container: {
    ...webStyles.container,
    minWidth: 1318,
    maxWidth: 1318
  }
}));

const mobileStyles: Styles = createStyles(({colors, spacing}) => ({
  ...webStyles,
  container: {
    ...webStyles.container,
    marginBottom: 40,
    minWidth: '88%',
    maxWidth: '88%'
  },
  bottomSection: {
    ...webStyles.bottomSection,
    marginRight: 124
  },
  button: {
    ...webStyles.button,
    width: 104,
    maxHeight: 41,
    minHeight: 41,
    marginRight: 20
  },
  chartContainer: {
    ...webStyles.chartContainer,
    marginLeft: 22,
    width: '116%'
  },
  rightSection: {
    ...webStyles.rightSection,
    marginRight: -52
  }
}));

export const AirQualityDrillDownContainer = () => {
  const dispatch = useDispatch();
  const {tabletMode, mobileMode} = useSelector((s: any) => s.styling);
  const dataRef = useRef({loadingSource: {isLoading: true}, chartData: null as any});
  const [activeView, setActiveView] = useState<string>('co');
  const [activeViewText, setActiveViewText] = useState<string>('Last 24 Hours CO Trendline');
  const {district} = useSelector((s: any) => s.map);
  const hourlyBuildingsCo = useFetch(fetchHourlyBuildingsCo, district);
  const hourlyBuildingsTemperature = useFetch(fetchHourlyBuildingsTemperature, district);
  const hourlyBuildingsHumidity = useFetch(fetchHourlyBuildingsHumidity, district);

  const styles = mobileMode ? mobileStyles : tabletMode ? tabletStyles : webStyles;

  useEffect(() => {
    switch (activeView) {
      case 'co':
        hourlyBuildingsCo.fetch(true);
        break;
      case 'humidity':
        hourlyBuildingsHumidity.fetch(true);
        break;
      case 'temperature':
        hourlyBuildingsTemperature.fetch(true);
        break;
    }
    return () => {
      hourlyBuildingsCo.isMountedRef.current = false;
      hourlyBuildingsTemperature.isMountedRef.current = false;
      hourlyBuildingsHumidity.isMountedRef.current = false;
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [district, activeView]);

  switch (activeView) {
    case 'co':
      dataRef.current.loadingSource = hourlyBuildingsCo;
      dataRef.current.chartData = {
        ...data,
        series: hourlyBuildingsCo.response?.chart ?? [],
        xAxis: {
          ...data.xAxis,
          categories: hourlyBuildingsCo.response?.categories ?? []
        }
      };
      break;
    case 'humidity':
      dataRef.current.loadingSource = hourlyBuildingsHumidity;
      dataRef.current.chartData = {
        ...data,
        series: hourlyBuildingsHumidity.response?.chart ?? [],
        xAxis: {
          ...data.xAxis,
          categories: hourlyBuildingsHumidity.response?.categories ?? []
        }
      };
      break;
    case 'temperature':
      dataRef.current.loadingSource = hourlyBuildingsTemperature;
      dataRef.current.chartData = {
        ...data,
        series: hourlyBuildingsTemperature.response?.chart ?? [],
        xAxis: {
          ...data.xAxis,
          categories: hourlyBuildingsTemperature.response?.categories ?? []
        }
      };
      break;
  }

  const switchToCo = () => {
    setActiveView('co');
    setActiveViewText('Last 24 Hours CO Trendline');
  };
  const switchToHumidity = () => {
    setActiveView('humidity');
    setActiveViewText('Last 24 Hours Humidity Trendline');
  };
  const switchToTemperature = () => {
    setActiveView('temperature');
    setActiveViewText('Last 24 Hours Temperature Trendline');
  };

  const closeContainer = useCallback(() => {
    dispatch(setIsAirQualityDrillDownOpened(false));
  }, [dispatch]);

  return (
    <div>
      <div style={styles.overlay} onClick={closeContainer}></div>
      <div className="animate__animated animate__fadeInUp" style={styles.container}>
        <div style={styles.bottomSection}>
          <div style={styles.switchesContainer}>
            <TextButton text="CO" style={styles.button} toggled={activeView === 'co'} onClick={switchToCo} />
            <TextButton text="HUMIDITY" style={styles.button} toggled={activeView === 'humidity'} onClick={switchToHumidity} />
            <TextButton text="TEMPERATURE" style={styles.button} toggled={activeView === 'temperature'} onClick={switchToTemperature} />
          </div>
          <div style={styles.rightSection}>
            {dataRef.current.loadingSource.isLoading ? (
              <Loader height="100px" margin={0} color={'#fff'} />
            ) : (
              <div style={styles.chartContainer}>
                <div style={styles.title}>
                  <Text variant="cardHeader">{activeViewText}</Text>
                </div>
                <HighchartsReact data-testid="highChart" highcharts={Highcharts} options={dataRef.current.chartData} />
              </div>
            )}
          </div>
        </div>
      </div>
    </div>
  );
};
