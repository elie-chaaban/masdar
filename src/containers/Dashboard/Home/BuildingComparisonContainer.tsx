import React, {useState, useCallback} from 'react';
import {Styles} from '../../../types';
import {createStyles} from '../../../theme';
import {useSelector, useDispatch, shallowEqual} from 'react-redux';
import {BuildingComparison, Building, BuildingInsightsDetails} from '../../../components/Dashboard/BuildingComparison';
import {Insight, Intensity, UnitFormat, unitFormatter, valueFormatter} from '../../../utils/dashboard';
// @ts-ignore
import {setIsBuildingComparisonOpened} from '../../../reduxStore/actions';
//@ts-ignore
import {fetchTotalEnergyConsumption, fetchCarbonEmissionAndTrendLine} from '../../../services/energy';
//@ts-ignore
import {fetchWaterDashboard, fetchWaterCarbonEmissionAndTrendLine} from '../../../services/water';
//@ts-ignore
import {fetchTotalWasteProduction, fetchWasteCarbonEmissionAndTrendLine} from '../../../services/waste';
//@ts-ignore
import {SlideIn} from '../../../components/Utils';
import {
  fetchRealTimeOperations,
  fetchFrequencyOfOccurence,
  fetchAcknowledgmentTime,
  fetchSecurityIncidentsCountBySeverity
  //@ts-ignore
} from '../../../services/security';

const styles: Styles = createStyles(({colors, spacing}) => ({
  container: {
    ...spacing.flex,
    position: 'absolute',
    maxHeight: 970,
    top: 0,
    left: 48,
    right: 0,
    bottom: 0,
    zIndex: 100002,
    backgroundColor: colors.buildingComparisonBackground,
    backdropFilter: 'blur(5.5px)'
  },
  containerIncreaseHeight: {
    maxHeight: 1100,
    top: 70
  },
  wasteDivertedValueWrapper: {
    display: 'flex',
    minHeight: 55,
    alignItems: 'flex-end'
  }
}));

export const BuildingComparisonContainer = (): JSX.Element => {
  const dispatch = useDispatch();
  const {buildings, wasteLocations, district} = useSelector((s: any) => s.map);
  const {selectedModule} = useSelector((s: any) => s.dashboard);
  const {filter, startDate, endDate} = useSelector((s: any) => s.dashboard.dateFilter, shallowEqual);
  const access = useSelector((s: any) => s.user.access.kpi);
  const hasEnergyAccess = access['energy']?.some((b: any) => b.district_id === district);
  const hasWaterAccess = access['water']?.some((b: any) => b.district_id === district);
  const hasCarbonAccess = access['carbon']?.some((b: any) => b.district_id === district);
  const hasWasteAccess = access['waste']?.some((b: any) => b.district_id === district);
  const accessModules = useSelector((s: any) => s.user.access.modules);
  const [buildingsInsights, setBuildingsInsights] = useState<BuildingInsightsDetails[]>([]);
  const [showLoading, setShowLoading] = useState(false);
  const [showInsights, setShowInsights] = useState(false);

  const prepareBuildings = useCallback((): Building[] => {
    const temp: Building[] = [];
    buildings.forEach((b: any) => {
      if (b.hasData) {
        if (b.imageFile?.filePath) {
          const imageUrl = `${process.env['REACT_APP_FILES_URL']}${b.imageFile.filePath}`;
          temp.push({...b, imageUrl});
        }
      }
    });
    return temp;
  }, [buildings]);

  const onDismiss = useCallback(() => {
    dispatch(setIsBuildingComparisonOpened(false));
  }, [dispatch]);

  const fetchSustainabilityBuildingInsights = useCallback(
    async (buildingId: string): Promise<BuildingInsightsDetails> => {
      const foundBuilding = buildings.find((b: any) => b.id === buildingId);
      const imageUrl = `${process.env['REACT_APP_FILES_URL']}${foundBuilding.imageFile.filePath}`;
      const foundWasteLocation = wasteLocations?.find((w: any) => w.buildingId === buildingId);
      const wasteLocation = foundWasteLocation?.location || '';

      let wasteProduction = {
        waste: 0,
        divertedPercentage: 0
      };

      if (wasteLocation && hasWasteAccess) {
        wasteProduction = await fetchTotalWasteProduction(filter, startDate, endDate, district, wasteLocation, buildingId);
      }

      const energyConsumption = hasEnergyAccess
        ? await fetchTotalEnergyConsumption(filter, startDate, endDate, district, buildingId)
        : {energy: 0, kwh: 0};
      const waterConsumption = hasWaterAccess
        ? await fetchWaterDashboard(filter, startDate, endDate, district, buildingId)
        : {water: 0, kwh: 0};
      const energyCarbonEmission =
        hasCarbonAccess & hasEnergyAccess
          ? await fetchCarbonEmissionAndTrendLine(filter, startDate, endDate, district, buildingId)
          : {value: 0};
      const waterCarbonEmission =
        hasCarbonAccess & hasWaterAccess
          ? await fetchWaterCarbonEmissionAndTrendLine(filter, startDate, endDate, district, buildingId)
          : {value: 0};
      const wasteCarbonEmission =
        hasCarbonAccess & hasWasteAccess
          ? await fetchWasteCarbonEmissionAndTrendLine(filter, startDate, endDate, district, buildingId)
          : {value: 0};

      const energy = [];

      if (hasEnergyAccess) {
        energy.push({
          id: 'energy-consumption',
          value: valueFormatter(Insight.ENERGY, energyConsumption?.energy, UnitFormat.String),
          unit: unitFormatter(Insight.ENERGY, energyConsumption?.energy)
        });
        energy.push({
          id: 'energy-intensity',
          value: valueFormatter(Intensity.ENERGY_INTENSITY, energyConsumption?.kwh, UnitFormat.String),
          unit: unitFormatter(Intensity.ENERGY_INTENSITY)
        });
        if (hasCarbonAccess)
          energy.push({
            id: 'energy-co2',
            value: valueFormatter(Insight.CARBON, energyCarbonEmission.value, UnitFormat.String),
            unit: 'CO2 ' + unitFormatter(Insight.CARBON, energyCarbonEmission?.value)
          });
      }

      const water = [];
      if (hasWaterAccess) {
        water.push({
          id: 'water-consumption',
          value: valueFormatter(Insight.WATER, waterConsumption?.water, UnitFormat.String),
          unit: unitFormatter(Insight.WATER)
        });
        water.push({
          id: 'water-intensity',
          value: valueFormatter(Intensity.WATER_INTENSITY, waterConsumption?.kwh, UnitFormat.String),
          unit: unitFormatter(Intensity.WATER_INTENSITY, waterConsumption?.kwh)
        });
        if (hasCarbonAccess)
          water.push({
            id: 'water-co2',
            value: valueFormatter(Insight.CARBON, waterCarbonEmission.value, UnitFormat.String),
            unit: 'CO2 ' + unitFormatter(Insight.CARBON, waterCarbonEmission?.value)
          });
      }

      const waste = [];
      if (hasWasteAccess) {
        waste.push({
          id: 'waste-production',
          value: valueFormatter(Insight.WASTE, wasteProduction?.waste, UnitFormat.String),
          unit: unitFormatter(Insight.WASTE, wasteProduction?.waste)
        });
        if (hasCarbonAccess) {
          waste.push({
            id: 'waste-co2',
            value: valueFormatter(Insight.CARBON, wasteCarbonEmission.value, UnitFormat.String),
            unit: 'CO2 ' + unitFormatter(Insight.CARBON, wasteCarbonEmission?.value)
          });
        }
        waste.push({
          id: 'waste-diverted',
          value: `${wasteProduction?.divertedPercentage} %`,
          unit: 'LANDFILL DIVERSION',
          valueWrapperCustomStyle: styles.wasteDivertedValueWrapper
        });
      }

      return {
        ...foundBuilding,
        imageUrl,
        energy,
        water,
        waste
      };
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [buildings, district, endDate, filter, startDate, wasteLocations]
  );

  const fetchSecurityBuildingInsights = useCallback(
    async (buildingId: string): Promise<BuildingInsightsDetails> => {
      const foundBuilding = buildings.find((b: any) => b.id === buildingId);
      const imageUrl = `${process.env['REACT_APP_FILES_URL']}${foundBuilding.imageFile.filePath}`;
      const realtimeOperationsData = await fetchRealTimeOperations(startDate, endDate, district, buildingId, null, filter);
      const securityIncidentsCountBySeverityData = await fetchSecurityIncidentsCountBySeverity(
        startDate,
        endDate,
        district,
        buildingId,
        null,
        filter
      );
      const frequencyOfOccurenceData = await fetchFrequencyOfOccurence(startDate, endDate, district, buildingId, null, filter);
      const acknowledgmentTimeData = await fetchAcknowledgmentTime(startDate, endDate, district, buildingId, null, filter);
      const realTimeOperations = [];
      realTimeOperations.push({
        id: 'real-time-open-operations',
        value: realtimeOperationsData?.openIncidentsChart?.total ?? 0,
        unit: 'OPEN'
      });
      realTimeOperations.push({
        id: 'real-time-completed-operations',
        value: realtimeOperationsData?.completedIncidentsChart?.total ?? 0,
        unit: 'COMPLETED'
      });

      const securityIncidentsCountBySeverity = [];
      securityIncidentsCountBySeverity.push({
        id: 'low-security-incidents-count',
        value: securityIncidentsCountBySeverityData?.low ?? 0,
        unit: 'LOW'
      });
      securityIncidentsCountBySeverity.push({
        id: 'medium-security-incidents-count',
        value: securityIncidentsCountBySeverityData.medium ?? 0,
        unit: 'MEDIUM'
      });
      securityIncidentsCountBySeverity.push({
        id: 'high-security-incidents-count',
        value: securityIncidentsCountBySeverityData.high ?? 0,
        unit: 'HIGH'
      });

      const acknowledgmentTime = [];
      acknowledgmentTime.push({
        id: 'low-acknowledgment-time',
        value:
          acknowledgmentTimeData?.acknowledgementTime_chartData
            ?.find((item: any) => item.name.toLowerCase() === 'low')
            .data.reduce((val: number, cur: number) => (val += cur), 0) ?? 0,
        unit: 'LOW'
      });
      acknowledgmentTime.push({
        id: 'medium-acknowledgment-time',
        value:
          acknowledgmentTimeData?.acknowledgementTime_chartData
            ?.find((item: any) => item.name.toLowerCase() === 'medium')
            .data.reduce((val: number, cur: number) => (val += cur), 0) ?? 0,
        unit: 'MEDIUM'
      });
      acknowledgmentTime.push({
        id: 'high-acknowledgment-time',
        value:
          acknowledgmentTimeData?.acknowledgementTime_chartData
            ?.find((item: any) => item.name.toLowerCase() === 'high')
            .data.reduce((val: number, cur: number) => (val += cur), 0) ?? 0,
        unit: 'HIGH'
      });

      const frequencyOfOccurence = [];
      frequencyOfOccurence.push({
        id: 'accessControl-frequency-of-occurence',
        value:
          frequencyOfOccurenceData?.trendLine
            ?.filter((x: any) => x.name === 'Access Control')[0]
            ?.data?.reduce((accumulator: number, currentValue: number) => accumulator + currentValue, 0) ??
          0 ??
          0,
        unit: 'ACCESS CONTROL'
      });
      frequencyOfOccurence.push({
        id: 'criminalIncident-frequency-of-occurence',
        value:
          frequencyOfOccurenceData?.trendLine
            ?.filter((x: any) => x.name === 'Criminal Incident')[0]
            ?.data?.reduce((accumulator: number, currentValue: number) => accumulator + currentValue, 0) ??
          0 ??
          0,
        unit: 'CRIMINAL INCIDENT'
      });
      frequencyOfOccurence.push({
        id: 'faultAlarms-frequency-of-occurence',
        value:
          frequencyOfOccurenceData?.trendLine
            ?.filter((x: any) => x.name === 'Fault Alarms')[0]
            ?.data?.reduce((accumulator: number, currentValue: number) => accumulator + currentValue, 0) ??
          0 ??
          0,
        unit: 'FAULT ALARMS'
      });
      frequencyOfOccurence.push({
        id: 'others-frequency-of-occurence',
        value:
          frequencyOfOccurenceData?.trendLine
            ?.filter((x: any) => x.name === 'Others')[0]
            ?.data?.reduce((accumulator: number, currentValue: number) => accumulator + currentValue, 0) ??
          0 ??
          0,
        unit: 'OTHERS'
      });
      frequencyOfOccurence.push({
        id: 'safety-frequency-of-occurence',
        value:
          frequencyOfOccurenceData?.trendLine
            ?.filter((x: any) => x.name === 'Safety')[0]
            ?.data?.reduce((accumulator: number, currentValue: number) => accumulator + currentValue, 0) ??
          0 ??
          0,
        unit: 'SAFETY'
      });

      return {
        ...foundBuilding,
        imageUrl,
        realTimeOperations,
        securityIncidentsCountBySeverity,
        frequencyOfOccurence,
        acknowledgmentTime
      };
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [buildings, district, endDate, filter, startDate, wasteLocations]
  );

  const onClickComparison = useCallback(
    async (buildingsIds: string[]) => {
      if (buildingsIds && buildingsIds.length) {
        setShowLoading(true);
        const temp: BuildingInsightsDetails[] = [];
        const promises = buildingsIds.map(async (buildingId) => {
          const buildingInsights =
            selectedModule === 'sustainability'
              ? await fetchSustainabilityBuildingInsights(buildingId)
              : await fetchSecurityBuildingInsights(buildingId);
          temp.push(buildingInsights);
        });
        await Promise.all(promises);
        setBuildingsInsights(temp);
        setShowLoading(false);
        setShowInsights(true);
      }
    },
    [selectedModule, fetchSustainabilityBuildingInsights, fetchSecurityBuildingInsights]
  );

  return (
    <SlideIn in animation={'slideUp'} style={{...styles.container, ...(accessModules?.length <= 1 ? styles.containerIncreaseHeight : {})}}>
      <BuildingComparison
        showLoading={showLoading}
        showInsights={showInsights}
        buildings={prepareBuildings()}
        buildingsInsights={buildingsInsights}
        onClickCancel={onDismiss}
        onClickClose={onDismiss}
        onClickComparison={onClickComparison}
      />
    </SlideIn>
  );
};
