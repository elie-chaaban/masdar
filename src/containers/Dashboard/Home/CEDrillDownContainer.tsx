import React, {useCallback, useEffect} from 'react';
import {Styles} from '../../../types';
import {createStyles} from '../../../theme';
import {useDispatch, useSelector, shallowEqual} from 'react-redux';
//@ts-ignore
import Loader from '../../../components/Utils/Loader';
import {
  FILTER_PV_CHART_INDEX,
  SustainabilityTrendLineAndBreakdown
} from '../../../components/Dashboard/SustainabilityTrendLineAndBreakdown';
// @ts-ignore
import {setDistrictPvProductionChart} from '../../../reduxStore/actions';
//@ts-ignore
import {fetchDistrictPVProductions} from '../../../services/energy';
//@ts-ignore
import useFetch from '../../../hooks';

const webStyles: Styles = createStyles(({colors, spacing}) => ({
  container: {
    position: 'absolute',
    bottom: 0,
    left: 48,
    minWidth: '100%',
    maxWidth: '100%',
    zIndex: 9999,
    color: colors.white
  },
  overlay: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    zIndex: 9999
  },
  bottomSection: {
    ...spacing.flexHorizontally,
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: colors.chartsCardBackground,
    gap: 20
  },
  pvProductionTrendLine: {
    maxWidth: 500
  }
}));

const tabletStyles: Styles = createStyles(({colors, spacing}) => ({
  ...webStyles,
  container: {
    ...webStyles.container,
    minWidth: 1318,
    maxWidth: 1318
  }
}));

const mobileStyles: Styles = createStyles(({colors, spacing}) => ({
  ...webStyles,
  container: {
    ...webStyles.container,
    left: 0,
    maxWidth: '98%',
    minWidth: '98%'
  },
  bottomSection: {
    ...webStyles.bottomSection,
    flexDirection: 'column',
    gap: 0
  }
}));

export const CEDrillDownContainer = () => {
  const dispatch = useDispatch();
  const {tabletMode, mobileMode} = useSelector((s: any) => s.styling);
  const styles = mobileMode ? mobileStyles : tabletMode ? tabletStyles : webStyles;
  const {district, buildings} = useSelector((s: any) => s.map, shallowEqual);
  const {districtPvProductionChart, dateFilter} = useSelector((s: any) => s.dashboard);
  const {filter} = dateFilter;

  const buildingPVProductions = useFetch(fetchDistrictPVProductions, district, '74894f39-bb41-73b0-09f2-3a0b82a7280c');

  useEffect(() => {
    if (district) {
      buildingPVProductions.fetch(true);
    }
    return () => {
      buildingPVProductions.isMountedRef.current = false;
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [district]);

  const closeContainer = useCallback(() => {
    dispatch(setDistrictPvProductionChart(null));
  }, [dispatch]);

  return (
    <div style={styles.overlay} onClick={closeContainer}>
      <div className="animate__animated animate__fadeInUp" style={styles.container}>
        <div style={styles.bottomSection}>
          {!districtPvProductionChart ? (
            <Loader height="100px" margin={0} color={'#fff'} />
          ) : (
            <SustainabilityTrendLineAndBreakdown
              variant="pvProductionTrendLine"
              style={styles.pvProductionTrendLine}
              series={districtPvProductionChart[FILTER_PV_CHART_INDEX[filter]]}
              categories={districtPvProductionChart[FILTER_PV_CHART_INDEX[filter]].categories}
              chartTitle="MASDAR 10 MW CLEAN ELECTRICITY"
            />
          )}
          {buildingPVProductions?.isLoading ? (
            <Loader height="100px" margin={0} color={'#fff'} />
          ) : (
            <SustainabilityTrendLineAndBreakdown
              variant="pvProductionTrendLine"
              style={styles.pvProductionTrendLine}
              series={buildingPVProductions?.response?.chart[FILTER_PV_CHART_INDEX[filter]]}
              categories={buildingPVProductions?.response?.chart[FILTER_PV_CHART_INDEX[filter]].categories}
              chartTitle="IRENA GENERATED CLEAN ELECTRICITY"
            />
          )}
        </div>
      </div>
    </div>
  );
};
