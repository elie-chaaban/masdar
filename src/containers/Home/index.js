import React, {useEffect} from 'react';
import {useDispatch, useSelector} from 'react-redux';
import Map from '../../components/Map';
import SideBar from '../SideBar';
import InsightSlider from '../../components/InsightSlider';
import IntersectionDataSlider from '../../components/IntersectionDataSlider';
import Slider from '../Slider';
import ActionDetails from '../ActionDetails';
import DynamicContent from '../DynamicContent';
import Modal from '../../components/Utils/Modal';
import MultiWindowModal from '../../components/Utils/MultiWindowModal';
import WindowNumberModal from '../../components/Utils/WindowNumberModal';
import MWBroadcastChannel from '../../services/MWBroadcastChannel';
import {LoaderFull} from '../../components/Utils';
import {setActiveWindows, setMultiWindowConfig, removeActiveWindow, toggleInsight, toggleSlider} from '../../reduxStore/actions';

const broadcastChannel = MWBroadcastChannel.getService();
broadcastChannel.startListening();

export default function Home() {
  const dispatch = useDispatch();
  const districts = useSelector((s) => s.map.districts);
  const districtsData = useSelector((s) => s.map.districtsData);
  const access = useSelector((s) => s.user.access);
  const doneLoading = districts && access && districtsData.length;

  useEffect(() => {
    dispatch(setMultiWindowConfig());
    dispatch(setActiveWindows());
    dispatch(toggleInsight('energy'));
    dispatch(toggleSlider());
    window.addEventListener('unload', (event) => {
      dispatch(removeActiveWindow());
    });
  }, [dispatch]);
  return doneLoading ? (
    <div className="App">
      <SideBar data-testid="sidebar" />
      <Map data-testid="map" />
      <Slider data-testid="slider" />
      <ActionDetails data-testid="actions" />
      <InsightSlider data-testid="insightSlider" />
      <IntersectionDataSlider data-testid="intersectionDataSlider" />
      <Modal data-testid="modal" />
      <MultiWindowModal data-testid="multiWindowModal" />
      <WindowNumberModal data-testid="windowNumberModal" />
      <DynamicContent data-testid="dynamicContent" />
    </div>
  ) : (
    <LoaderFull />
  );
}
