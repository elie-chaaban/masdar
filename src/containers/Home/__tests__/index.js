import React from 'react';
import {Provider} from 'react-redux';
import {act} from 'react-dom/test-utils';
import {mount} from 'enzyme';
import {mockStore} from '../../../testUtils';
import Home from '..';

jest.mock('../../../components/Map', () => ({
  __esModule: true,
  default: () => <div />
}));
jest.mock('../../Dashboard', () => ({
  __esModule: true,
  default: () => <div />
}));
jest.mock('../../Slider', () => ({
  __esModule: true,
  default: () => <div />
}));
jest.mock('../../ActionDetails', () => ({
  __esModule: true,
  default: () => <div />
}));
jest.mock('../../SideBar', () => ({
  __esModule: true,
  default: () => <div />
}));
jest.mock('../../../components/Utils/Modal', () => ({
  __esModule: true,
  default: () => <div />
}));
jest.mock('../../../components/Utils/MultiWindowModal', () => ({
  __esModule: true,
  default: () => <div />
}));
jest.mock('../../../components/Utils/WindowNumberModal', () => ({
  __esModule: true,
  default: () => <div />
}));
jest.mock('../../DynamicContent', () => ({
  __esModule: true,
  default: () => <div />
}));
jest.mock('../../../components/IntersectionDataSlider', () => ({
  __esModule: true,
  default: () => <div />
}));
test('Home component', async () => {
  const state = () => ({
    dashboard: {
      isMultiWindowModalOpened: false,
      multiWindowModalMode: 'normal',
      dynamicContentType: null,
      activeWindowAction: null,
      activeWindowId: null,
      activeWindows: [],
      multiWindowConfig: null,
      dateFilter: {
        filter: 'year',
        startDate: '2020-05-01',
        endDate: '2020-09-01'
      },
      isOpenInsightSlider: false
    },
    map: {
      district: '1',
      districtInfo: {
        name: 'telus',
        location: 'canada',
        longitude: 43,
        latitude: 20
      }
    },
    user: {
      access: {
        insights: ['1']
      }
    }
  });
  const store = mockStore(state());
  let dom;
  await act(async () => {
    dom = mount(
      <Provider store={store}>
        <Home />
      </Provider>
    );
  });
  dom.update();
  expect(dom).toMatchSnapshot();
  expect(dom.find("[data-testid='sidebar']")).toHaveLength(1);
  expect(dom.find("[data-testid='map']")).toHaveLength(1);
  expect(dom.find("[data-testid='dashboard']")).toHaveLength(1);
  expect(dom.find("[data-testid='slider']")).toHaveLength(1);
  expect(dom.find("[data-testid='actions']")).toHaveLength(1);
  expect(dom.find("[data-testid='insightSlider']")).toHaveLength(1);
  expect(dom.find("[data-testid='modal']")).toHaveLength(1);
  expect(dom.find("[data-testid='multiWindowModal']")).toHaveLength(1);
  expect(dom.find("[data-testid='windowNumberModal']")).toHaveLength(1);
  expect(dom.find("[data-testid='dynamicContent']")).toHaveLength(1);
});
