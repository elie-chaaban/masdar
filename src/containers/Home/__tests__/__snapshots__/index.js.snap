// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Home>
    <div
      className="App"
    >
      <default
        data-testid="sidebar"
      >
        <div />
      </default>
      <default
        data-testid="map"
      >
        <div />
      </default>
      <default
        data-testid="dashboard"
      >
        <div />
      </default>
      <default
        data-testid="slider"
      >
        <div />
      </default>
      <default
        data-testid="actions"
      >
        <div />
      </default>
      <InsightSlider
        data-testid="insightSlider"
      >
        <SlideIn
          className="slider-details-container dark-container insight-slider "
          in={false}
        >
          <Animate
            animation="fadeRight"
            className="slider-details-container dark-container insight-slider "
            enter={false}
            onAnimationEnd={null}
            options={
              Object {
                "duration": 500,
              }
            }
            unMountOnExit={true}
          />
        </SlideIn>
      </InsightSlider>
      <default
        data-testid="intersectionDataSlider"
      >
        <div />
      </default>
      <default
        data-testid="modal"
      >
        <div />
      </default>
      <default
        data-testid="multiWindowModal"
      >
        <div />
      </default>
      <default
        data-testid="windowNumberModal"
      >
        <div />
      </default>
      <default
        data-testid="dynamicContent"
      >
        <div />
      </default>
    </div>
  </Home>
</Provider>
`;
